src/transforms/projectionpursuit.jl

```diff
@@ -0,0 +1,205 @@
+# ------------------------------------------------------------------
+# Licensed under the MIT License. See LICENSE in the project root.
+# ------------------------------------------------------------------
+
+"""
+    ProjectionPursuit(;tol=1e-6, maxiter=100, deg=5, perc=.9, n=100)
+
+The projection pursuit multivariate transform converts any multivariate distribution into
+the standard multivariate Gaussian distribution.
+
+This iterative algorithm repeatedly finds a direction of projection `α` that maximizes a score of
+non-Gaussianity known as the projection index `I(α)`. The samples projected along `α` are then
+transformed with the [`Quantile`](@ref) transform to remove the non-Gaussian structure. The
+other coordinates in the rotated orthonormal basis `Q = [α ...]` are left untouched.
+
+The non-singularity of Q is controlled by assuring that norm(det(Q)) ≥ `tol`. The iterative 
+process terminates whenever the transformed samples are "more Gaussian" than `perc`% of `n`
+randomly generated samples from the standard multivariate Gaussian distribution, or when the 
+number of iterations reaches a maximum `maxiter`.
+
+# Examples
+
+```julia
+ProjectionPursuit()
+ProjectionPursuit(deg=10)
+ProjectionPursuit(perc=.85, n=50)
+ProjectionPursuit(tol=1e-4, maxiter=250, deg=5, perc=.95, n=100)
+```
+
+"""
+
+struct ProjectionPursuit{T} <: Stateless
+  tol::T
+  maxiter::Int
+  deg::Int
+  perc::T
+  n::Int
+end
+
+ProjectionPursuit(;tol=1e-6, maxiter=100, deg=5, perc=.95, n=100) =
+  ProjectionPursuit{typeof(tol)}(tol, maxiter, deg, perc, n)
+
+isrevertible(::Type{<:ProjectionPursuit}) = true
+
+# transforms a row of random variables into a convex combination 
+# of random variables with values in [-1,1] and standard normal distribution
+function rscore(Z, α)
+  ᾱ = (1/norm(α)) .* α
+  X = Z * ᾱ
+  2 .* cdf.(Normal(), X) .- 1
+end
+
+# projection index along a given direction
+function pindex(transform, Z, α)
+  X = rscore(Z, α)
+  I = (3/2) * mean(X)^2
+  if transform.deg > 1
+    Pⱼ₋₂, Pⱼ₋₁ = ones(size(X)), X
+    for j = 2:transform.deg
+      Pⱼ₋₂, Pⱼ₋₁ = 
+        Pⱼ₋₁, (1/j) * ((2j-1) * X .* Pⱼ₋₁ - (j-1) * Pⱼ₋₂)
+      I += ((2j+1)/2) * (mean(Pⱼ₋₁))^2
+    end
+  end
+  I
+end
+
+# find axis along which projection index is maximum
+function optimaxis(transform, Z)
+  _, q = size(Z)
+  E = 1.0 * I(q)
+  proj(i) = pindex(transform, Z, E[:,i])
+  Iₘₐₓ, iₘₐₓ = findmax(proj, 1:q)
+  E[:,iₘₐₓ]
+end
+
+function optimdiag(transform, Z, ᾱ)
+  q  = length(ᾱ)
+  Iproy₀ = pindex(transform, Z, ᾱ)
+  Id = 1. *I(q)
+  for i in 1:q
+    eᵢ = Id[:,i]
+    f₊ = pindex(transform, Z, (1/√(2+2ᾱ[i])) * (ᾱ + eᵢ))
+    f₋ = pindex(transform, Z, (1/√(2-2ᾱ[i])) * (ᾱ - eᵢ)) * (ᾱ[i] != 1)
+    f = (f₊ > f₋) * f₊ + (f₊ ≤ f₋) * f₋ 
+    s = (f₊ > f₋) * 1 + (f₊ ≤ f₋) * (-1)
+    if f > pindex(transform, Z, ᾱ)
+      ᾱ = (1/√(2+2s*ᾱ[i])) * (ᾱ + s*eᵢ)
+    end
+  end
+  pindex(transform, Z, ᾱ) ≥ Iproy₀ ? ᾱ : optimaxis(transform, Z)
```

Comment:
I think the logic in this function can be simplified. Can we start our iteration assuming that the input alpha is already the axis with maximum score? Then, we just need to loop over all diagonals and check if any of them has a better score than the input. If any has better score, we update alpha and the score.

The algorithm seems correct overall, just a bit confusing with a call to optimaxis at the end. The pseudo code should read as

```julia
# input as first guess
α = ᾱ
I = pindex(transform, Z, ᾱ)

# check diagonals
for i in 1:q
  # compute pindex along diagonals
  αd = ...
  Id = ...
  # check and update
  if Id > I
    α = αd
    I = Id
  end
end
# now we have the updated alpha
α